Software: Socket.io (#21)
PoCInnovation/Workshops: software/19.Socket.io/README.md

# Workshop 19 - Chat with Socket.IO

> ✔ Understand the sockets

Reviewer: That the only thing you gonna learn in this workshop? :(

> ## Step 0 : Setup

Reviewer: ```suggestion
## Step 0 - Setup
```

Any information about the setting of this workshop are available in [SETUP.md](./SETUP.md).

> Be careful to finish every step before moving one to the next thing.

Reviewer: Fix the english.

You can use that sentence for example
```
All the required information to start this workshop can be found in [SETUP.md](./SETUP.md)
```

> A cheatsheet based on Typescript is available [here](https://github.com/PoCInnovation/Workshops/blob/master/software/8.Typescript/HELP.md).

Reviewer: Don't use `here`, just add the link like below

```suggestion
If you are new to Typescript, you should check our [Typescript cheatsheet](https://github.com/PoCInnovation/Workshops/blob/master/software/8.Typescript/HELP.md).
```

[What's a socket ?](https://socket.io/docs/v4/index.html)

> ## Step 01 :  First step, Hello World !

Reviewer: ```suggestion
## Step 01 -  First step, Hello World !
```

If the installation has been managed correctly, you can run your server with
```bash
 npm run dev
```

If you go on your browser at http://localhost:8080/
you should see this on the screen :

```
Hello World !
```
The server launch and don't need to be launch again with nodemon.
The script in your package.json allow to take account of modifications

> of your code to launch the server by itself.

Reviewer: Should be in the setup, you just verify that everything has been successfully setup

> ## Step 02 :

Reviewer: :warning: Replace `:` with `-` for each steps and add titles

> Your server work with express, you can find it in

Reviewer: Add link to express

> src/index.ts.

Reviewer: Add link
E.g
```suggestion
[src/index.ts](./src/index.ts).
```

> You need to change this file to allow express and socket.io to use it at the same time.

Reviewer: Add links to express and socket.io
```suggestion
You need to change this file to both use express and socket.io.
```

That will allow you to issue and receive data on your server.

> 

Reviewer: OK but what we should do ??? Should we need to add something somewhere ? Do we need to create a new serveur ? Is there a middleware to apply ???

> Once the server is set, we will determine the path of our html page on our route '/' that you can find in router/router.ts.

Reviewer: Link to file.
Where is the HTML page? Should we define a new resolver in our express server? Which method?

> To every connection or disconnection of a user, we will send a message to our shell with the help of console.log().

Reviewer: What kind of message? What is the name of the function to write? Where should we add it?

#### Ressources :
- [Server Socket.IO](https://socket.io/get-started/chat)
- [Introduction Socket.IO](https://socket.io/docs/v4/index.html)
- [Socket.IO](https://socket.io/)
- [Link HTML](https://www.techiediaries.com/express-sendfile-serve-static-files/)

> ## Step 03 :

Reviewer: Replace `:` by `-` and add title

> Creat a class in the file Front/chat.ts that will allow us to receive and send messages in the form of socket to your server.

Reviewer: Link to file.
Which methods should we define? Do they take arguments? What are their return types? They are async?

> Star with the famous 'Hello World' that you will display in your shell.

Reviewer: Replace `'` with backtick 
E.g
```
`Hello world`
```

#### Ressources :
- [Class](https://www.typescriptlang.org/docs/handbook/2/classes.html)
- [Base Affichage](https://www.typescriptlang.org/docs/handbook/typescript-from-scratch.html)

> ## Step 04 :

Reviewer: Replace `:` by `-` and add title

> It is time to get information on your input.

Reviewer: ```suggestion
It's time to get information on your input.
```

> For that you need to create a function that achieve a query on the id of your "input" once this one is submit. 

Reviewer: Ok but what is the function's name? Arguments? Return type? 
Replace `"` with backtick

> Once you get the message, you need to get the message of your server to print it on your shell.

Reviewer: How do we do that? I think you can add details

#### Ressources :
- [Socket listener](https://socket.io/docs/v4/index.html)

> ## Step 05 :

Reviewer: Replace `:` by `-` and add title

Having our message on your shell is great but receive your message in your browser console is better.

Our message from input are, for now, send to our server; We need to aks it to retransmit it with the aim of print it in the browser console.

> Creat a function who get back the message nearby the server to print it in the browser console.

Reviewer: Add details and fix your english please

#### Ressources :
- [Socket Client](https://socket.io/docs/v4/client-socket-instance/)

> ## Step 06 :

Reviewer: Replace `:` by `-` and add title

With this step, you will send a message to all the user on the page HTML each time there is a connection or disconnection.

For this end, we will creat a function that add contents with the help of a query on our HTML page.
You have to take in consideration the message that your server display to sent it to your front.

To test your code, we will open two pages of your browser with the url : http://localhost:8080/

Throughout the connection, you have to print "Connection" and "Disconnexion" when you leave the page.

> 

Reviewer: Add details and fix english typos

#### Ressources :

- [Emettre une socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)
- [Requête HTML](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector)

> ## Step 07 :

Reviewer: Replace `:` by `-` and add title

Creat a function that sent the content of our input on our HTML page.

For this, we will create a function that achieve a query on our HTML page to add it the message of the input.

#### Ressources :

- [Release a socket] (https://socket.io/docs/v3/emit-cheatsheet/index.html)
- [Request HTML] (https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector)
- [Socket Client](https://socket.io/docs/v4/client-socket-instance/)

> ## Step 08 :

Reviewer: Replace `:` by `-` and add title

It is time to identify the message more clearly. 
Creat an id for every user to allow in order to receive messages of users with their name or a determinant.

Exemple:

You: "message_sent"

> Tony: "message_receive"

Reviewer: Add details and fix typos

> ## Step 09 :

Reviewer: Replace `:` by `-` and add title

For this last step we need tro add the date and the hour of when the message was sent.

## Bonus

* Creat rooms that allow you to send a message to every user that are in the same room as you.
* Amelioration of your Front with React.
* Authentification of the user with socket.

> * Add a data base to stock the message et refresh when there is a connexion.

Reviewer: Typo

#### Ressources:
- [React](https://reactjs.org/)
- [Room Socket](https://socket.io/docs/v3/rooms/index.html)

> - [Authentifaciton Socket](https://socket.io/docs/v3/middlewares/)

Reviewer: You missed authors and organization.
See [other workshops](https://github.com/PoCInnovation/Workshops/tree/master/software/) for details

After that, print `New user + name` with username, when a user connects to the page.

#### Resources :
[Set Name Variable](https://www.dummies.com/web-design-development/javascript/how-to-prompt-the-user-for-input-in-javascript/)

## Step 10 - Message Management

Emit event `new-user` with `my_name` like parameter in chat.ts to prevent the arrival
of a new user to the server.
Receive message in your server add name in array variable `username` and
print in your console `new user` with her name.

In your chat.ts add the name of your client before the message.
You must change the print value of when you add your div.

When you log in, you have `New User` + my_name and` name: message` with each message sent.

Congratulation you have a chat online !

#### Resources :

## Bonus

* Set connection and disconnect.
* Create rooms that allow you to send a message to every user that are in the same room as you.
* Amelioration of your Front with React.
* Authentication of the user with socket.
* Add a database to stock the message et refresh when there is a connexion.

#### Resources :
- [React](https://reactjs.org/)
- [Room Socket](https://socket.io/docs/v3/rooms/index.html)

> - [Authentication Socket](https://socket.io/docs/v3/middlewares/)

Reviewer: Your forgot authors & organization

#### Resources :
- [Introduction Socket.IO](https://socket.io/docs/v4/index.html)
- [Socket.IO](https://socket.io/)

## Step 04 - Transmission Data

You have established connection !

It's time to implement a function to send messages from your frontend
that will be received on your server.

Now, go to the file `src/front/chat.ts` and complete the Class `Chat`:
* Create a function `emit_data` with for parameter a string.
* Emit socket with parameter `data` and your message.
* In your index.ts you must get information with your event `data' and the message.
* Tips: read the resource `Server Socket` and how receive your socket after emit it.
* Call your function in the `chat.ts` set `Hello World !` in the parameter.

In your terminal you should have to display your `Hello World !`.

#### Resources :
- [Class](https://www.typescriptlang.org/docs/handbook/2/classes.html)
- [Base Print](https://www.typescriptlang.org/docs/handbook/typescript-from-scratch.html)
- [Emit Socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)

## Step 05 - Get Browser Data

Now that we will get the data, we will send it from our server in index.ts.

> 

Reviewer: Add a schema to explain how it gonna works

> Now that we will get the data, we will send it from our server in index.ts.
> 
> The objective of this step is to know when the user presses the button on your html page.
> At this point you will add the content of the input into your terminal.
> 
> It is time to get information on your input.
> 
> First step: create two const variable:

Reviewer: ```suggestion
Now that we retrieved the data, we will send it from our server to our fronted.

The objective of this step is to know when the user presses the button on the html page.
At this point you will log the content of the input into your terminal.

It's time to get information from your input.

##### Retrieve html input components

In `chat.ts`, you will create two constants variables:
```

> * form_data: get element form by id on your html page.
> * input_data: get element input by id on your html page.
> 
> Second step:
> * Listen to the event submit button.
> * Create a `message` variable that gets the data from your input.
> * Print content message with console.log().
> * Call emit_data to sent `message`.
> * Established input in Null.
> * Tips: to get value of your data_input cast your variable with `HTMLInputElement`.

Reviewer: ```suggestion
- `form_data`: get element `form` by his id on your html page.
- `input_data`: get element `input` by his id on your html page.

##### Send message
- Make your button listen to the event `submit` .
- Create a `message` variable to store the data from your input.
- Log the content.
- Call `emit_data` to send the `message`.
- Update input's value to an empty string.

> :bulb: To get value from `data_input`, cast your variable to `HTMLInputElement`.
>
> :warning: It's necessary to always update the input value with an empty string to doesn't keep the previous value. 
```

> Go to your console on your shell and watch if you receive the message when you
> click on the button `Send`.

Reviewer: ```suggestion
You should see the message logged in your terminal after you have clicked the button `Send`
```

> #### Resources :

Reviewer: ```suggestion
#### Resources
```

- [Socket listener](https://socket.io/docs/v4/index.html)
- [Cast Variable](https://www.typescripttutorial.net/typescript-tutorial/type-casting/)

## Step 06 - Print Data in Your Browser Console

> Having our message on your shell is great but receive your message in your browser console is better.
> 
> In your server `index.ts` changed the sending of sockets.
> Make sure that everyone can receive the message.
> 
> * Change your socket.emit of your file `index.ts`for send socket to all clients in the current namespace except the sender.
> 
> Now, if you submit message and look your console browser you must see your message.
> 
> #### Resources :
> - [Socket Client](https://socket.io/docs/v4/client-socket-instance/)
> - [Type of emit Socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)

Reviewer: ```suggestion
See our message on our shell is cool but see our message in our browser's console is better.

In your backend, change the way of sockets' sending to send the received message to **everyone**.

- Change your `socket.emit` in `index.ts` to send `socket` to all clients in the current namespace __except to the sender__.

Now, if you submit a message and look to the browser's console: you should see your message.

#### Resources
- [Socket Client](https://socket.io/docs/v4/client-socket-instance/)
- [Type of emit Socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)
```

## Step 07 - Print Data in your Browser Html Page

> Now we will create a function in our `Chat` class to add the content of our input to our div.
> This will allow us to see the messages directly in our HTML page.
> 
> For this, we will create a function `append_data` on your class Chat.
> First step, create const variable:
> * html_page: get element by id of your `div`, the name of this id is `messages'.
> 
> Your second step is in your function `append_data`:
> * Create variable `add_element` like element  `div`.
> * The attribute the value of this variable to the message of your parameter.
> * Append this variable on your variable `html_page`.
> 
> Finally, call this function in the function `emit_data`.
> 
> If you send a message, you must see in your HTML page !!!!!
> Congratulation !
> 
> #### Resources :
> 
> - [Release a socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)
> - [Request HTML](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector)
> - [Socket Client](https://socket.io/docs/v4/client-socket-instance/)

Reviewer: ```suggestion
Now we will create a new method in our `Chat` class to add the content of our input into our div.
This will allow us to see the messages directly in our HTML page.

To do this, we will create a method `append_data` in our class `Chat`.
First, create a constant variable:
- `html_page`: get element by his id of you `div`, his id should be is `messages'.

Second, in your method `append_data`:
- Create a variable `add_element` that should create an element `div`.
- Display the message gave as parameter in this new div.
- Append the `div` to your variable `html_page`.

Finally, call this method in the `emit_data`.

If you send a message, you should see it in your HTML page !!!
Congratulation !

#### Resources
- [Release a socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)
- [Request HTML](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector)
- [Socket Client](https://socket.io/docs/v4/client-socket-instance/)
```

> ## Step 08 - Set username
> 
> We will now retrieve the name of each user when they connect.
> 
> we will:
> * Create a `my_name` variable that retrieves a user's name through a prompt that says `What is your name?`.
> * Just bellow, print `New user + name` with username, when a user connects to the page.
> 
> #### Resources :
> [Set Name Variable](https://www.dummies.com/web-design-development/javascript/how-to-prompt-the-user-for-input-in-javascript/)

Reviewer: ```suggestion
## Step 08 - Get username

We will now retrieve the name of each user when they connect.

You will need to:
- Create a variable `my_name` that retrieves the username through a [prompt](https://wprock.fr/blog/javascript-boites-dialogue-alert-confirm-prompt/) that says `What is your name?`.
- Just bellow, display `New user + name` with the username, when a user connects to the page.

#### Resources
[Set Name Variable](https://www.dummies.com/web-design-development/javascript/how-to-prompt-the-user-for-input-in-javascript/)
```

> ## Step 09 - Set Username
> 
> We are going to emit a socket that defines when a user connects with as parameter his name.
> This event will be sent to our server which in turn will send a socket to our frontend to ask
> it to display a connection message.
> 
> In chat.ts, we will:
> * Create event `new_user` with `my_name` like parameter in chat.ts to prevent the arrival
> of a new user to the server.
> * Create listen socket `user_connect` with string like parameter.
> * Inside, call `emit_data` with string `Connect + name`.
> 
> In the index.ts we will:`
> * Create listen socket `new_user` with string like parameter.
> * Inside, print in your terminal `New user + name`.
> * emit socket to all user with `user-connect` as event and name as parameter.
> 
> In your chat.ts add the name of your client before the message.
> You must change the print value of when you add your div.
> 
> When you log in, you have `New User` + my_name.
> 
> #### Resources :

Reviewer: ```suggestion
## Step 09 - Set Username

We are going to emit a socket triggered when a user connects with his name as parameter.<br>
This event will be sent to our server which will send back a socket to our frontend to display a connection message.

In `chat.ts`, we will:
- Create an event `new_user` with `my_name` as parameter.
- Create a listening socket named `user_connect` with a `string` as a parameter.
- Inside the socket, call `emit_data` with the string `Connect + name` where `name` is the parameter.

In the `index.ts` we will:
- Create a listening socket named `new_user` with a `string` as a parameter.
- Inside it, display in your terminal `New user + name` where `name` is the parameter.
- emit socket to all users with an `user-connect` event and `name` as parameters.

In `chat.ts`, add the name of your client before the message.
You must change the value of the print when you add a div.

When you log in, you should see `New User + your_name`.
```

> ## Step - Manage Message
> 
> The final touch is to add the name of each user before the message sent.
> 
> In your chat.ts add the name of your client before the message.
> * Change the value of your display when you add the content of your input to your div.
> Example: `name: message`.
> 
> If you send messages on your HTML page you must have the user's name before the message.
> 
> Congratulation you have a chat online !

Reviewer: ```suggestion
## Step 10 - Manage Message

The final touch is to add the name of each user before the message is sent.

In `chat.ts`, add the name of the client before the message.
- Change the value of the display when you add the content of the input into your div.

:bulb: **Example**: `name: message`.

If you send messages on your HTML page you should see the username before the message.

Congratulation you have an online chat !
```

## Bonus

> * Set connection and disconnect.
> * Create rooms that allow you to send a message to every user that are in the same room as you.
> * Amelioration of your Front with React.
> * Authentication of the user with socket.
> * Add a database to stock the message et refresh when there is a connexion.

Reviewer: ```suggestion
* Set connection and disconnect.
* Create rooms that allow you to send a message to every user that is in the same room.
* Improve your fronted with React.
* Authentication of the user with socket.
* Add a database to store messages and refresh when there is a connection.
```

> #### Resources :

Reviewer: ```suggestion
#### Resources
```

# Workshop 19 - Chat with Socket.IO

### Learn Socket.IO by implementing your own online chat.

> You will : <br/>

Reviewer: Remove those two lines

:heavy_check_mark: Understand sockets<br/>
:heavy_check_mark: Understand how transmits data<br/>

## Story

> The schem below shows an online chat:<br/>
> ![Data scheme](../../.github/assets/software/Socket.IO/Untitled-2021-08-23-0042.png) <br/>
> **Explanation**
> Each client represents an HTML page.<br/>
> As soon as a client sends a message, it is transmitted as a socket to the backend.<br/>
> The backend retrieves the socket and sends it to all clients as a socket.<br/>
> The content of the message is then displayed on the HTML page of all clients.<br/>
> We will then display the message on our frontend.<br/>

Reviewer: ```suggestion
The schema below shows the architecture of the online chat:
![Data scheme](../../.github/assets/software/Socket.IO/Untitled-2021-08-23-0042.png)

**Explanation**
Each client represents an HTML page.<br/>
As soon as a client sends a message, it is transmitted as a socket to the backend.<br/>
The backend retrieves the socket and sends it to all clients as a socket.<br/>
The content of the message is then displayed on the HTML page of all clients.<br/>
We will then display the message on our frontend.<br/>
```

> The schem bellow shows an online chat:<br/>
> ![Data scheme](../../.github/assets/software/Socket.IO/Untitled-2021-08-23-0042.png) <br/>
> **Explanation**
> Each client represents an HTML page.<br/>
> As soon as a client sends a message, it is transmitted as a socket to the backend.<br/>
> The backend sends retrieves the socket and sends it to all clients as a socket.<br/>
> The content of the message is then transferred to the HTML page of all clients.<br/>

Reviewer: ```suggestion
The schem below shows an online chat:<br/>
![Data scheme](../../.github/assets/software/Socket.IO/Untitled-2021-08-23-0042.png) <br/>
**Explanation**
Each client represents an HTML page.<br/>
As soon as a client sends a message, it is transmitted as a socket to the backend.<br/>
The backend retrieves the socket and sends it to all clients as a socket.<br/>
The content of the message is then displayed on the HTML page of all clients.<br/>
```

> Nous afficherons ensuite le message sur notre frontend.<br/>

Reviewer: Nice english haha

## Step 0 - Setup

All the required information to start this workshop can be found in [SETUP.md](./SETUP.md).

Be careful to finish every step before moving one to the next thing.

If you are new to Typescript, you should check our [Typescript cheatsheet](https://github.com/PoCInnovation/Workshops/blob/master/software/8.Typescript/HELP.md).

## Step 01 - Link your HTML page

Your first step is to modify the content of your '/' routes
in order to have your html page on your route.

In the file `src/router/router.ts`
- import [path](https://nodejs.org/docs/latest/api/path.html)
- replace `Hello World` by your HTML page's path define in `src/Front/index.html`.

> If your reload your browser page, you must see input and button `Send`.

Reviewer: ```suggestion
If your reload your browser page, you should see the input and button `Send`.
```

The backend retrieves the socket and sends it to all clients as a socket.<br/>
The content of the message is then displayed on the HTML page of all clients.<br/>
We will then display the message on our frontend.<br/>

- [What's a socket ?](https://socket.io/docs/v4/index.html)

## Step 0 - Setup

All the required information to start this workshop can be found in [SETUP.md](./SETUP.md).

Be careful to finish every step before moving one to the next thing.

If you are new to Typescript, you should check our [Typescript cheatsheet](https://github.com/PoCInnovation/Workshops/blob/master/software/8.Typescript/HELP.md).

## Step 01 - Link your HTML page

Your first step is to modify the content of your '/' routes
in order to have your html page on your route.

In the file `src/router/router.ts`
- import [path](https://nodejs.org/docs/latest/api/path.html)
- replace `Hello World` by your HTML page's path define in `src/Front/index.html`.

If your reload your browser page, you should see the input and button `Send`.

#### Resources
- [Set path](https://www.codegrepper.com/code-examples/html/express+sendFile)

## Step 02 - Create Socket Server

> First, you can find an  [express](https://expressjs.com/fr/) server in `src/index.ts`.
> You need to change this file to both use express and socket.io.
> That will allow you to issue and receive data on your server.

Reviewer: ```suggestion
First, you can find an  [express](https://expressjs.com/fr/) server in `src/index.ts`.<br/>
You need to change this file to both use express and socket.io.<br/>
That will allow you to issue and receive data on your server.
```

- [Introduction Socket.IO](https://socket.io/docs/v4/index.html)
- [Socket.IO](https://socket.io/)

## Step 04 - Transmission Data

You have established connection !

It's time to implement a method to send messages from your frontend to your backend.

Now, complete the Class `Chat` in `src/front/chat.ts`  :
- Create a method `emit_data` with a `message` of type `string` as parameter.
- Emit socket with your event named `data` and your `message`.
- In `index.ts` you must receive the information from the `data` event log the `message`.
- Call your `emit method` in `chat.ts` with `Hello World !` as message.

> :bulb: Take your time to read the documentation.

In your terminal you should see `Hello World !` logged.

#### Resources
- [Typescript Class](https://www.typescriptlang.org/docs/handbook/2/classes.html)
- [Log message](https://www.typescriptlang.org/docs/handbook/typescript-from-scratch.html)
- [Emit Socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)
- [Socket Server](https://k6.io/docs/javascript-api/k6-ws/socket/socket-on-event-callback/)

## Step 05 - Get Browser Data

Now that we retrieved the data, we will send it from our server to our fronted.

> The objective of this step is to know when the user presses the button on the html page.
> At this point you will log the content of the input into your terminal.

Reviewer: ```suggestion
The objective of this step is to know when the user presses the button on the html page.<br/>
At this point you will log the content of the input into your terminal.
```

#### Resources
- [Socket listener](https://socket.io/docs/v4/index.html)
- [Cast Variable](https://www.typescripttutorial.net/typescript-tutorial/type-casting/)

## Step 06 - Print Data in Your Browser Console

See our message on our shell is cool but see our message in our browser's console is better.

In your backend, change the way of sockets' sending to send the received message to **everyone**.

- Change your `socket.emit` in `index.ts` to send `socket` to all clients in the current namespace __except to the sender__.

Now, if you submit a message and look to the browser's console: you should see your message.

#### Resources
- [Socket Client](https://socket.io/docs/v4/client-socket-instance/)
- [Type of emit Socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)

## Step 07 - Print Data in your Browser Html Page

Now we will create a new method in our `Chat` class to add the content of our input into our div.
This will allow us to see the messages directly in our HTML page.

To do this, we will create a method `append_data` in our class `Chat`.
First, create a constant variable:

> - `html_page`: get element by his id of you `div`, his id should be is `messages'.

Reviewer: ```suggestion
- `html_page`: get element by his id of you `div`, his id should be is `messages`.
```
